### PR TITLE
[Features] Enable partitioned variable assignments

### PR DESCRIPTION
**Pull request by wangsiyu**
`PartitionedVariable` cannot use `assign`, `assign_add` and `assign_sub` functions in graph mode. In this pull request, I have enabled these functions in graph mode. This is the first step to enable variables partitioner usage in distribution strategy.

### CODE REVIEW

**Comment by wangsiyu** [comment]
@ebrevdo @theweiho Any coding style and design review comments are welcome. @yuefengz This maybe the first step to enable variable partitioned usage in distribution strategy

**Comment by yangjunpro** [comment]
Thank @wangsiyu for the work. 
@yuefengz  @ebrevdo @theweiho any feedback from your sides are highly appreciated.

Talk is cheap, show the code.

Siyu directly made this PR to ensure the communication could be more efficient.

**Comment by wangsiyu** [comment]
Unrelated failures? @ebrevdo @theweiho I noticed master branch also failed with these unitest especially in  XLA.

**Comment by facaiy** [comment]
@alextp Hi, Alexandre. The PR seems to be duplicated with tf 2.0 tensorflow/community#11 , right?

> symbols like tf.assign* will be removed in favor of methods in tf.Variable

Could you make a comment or reassign? Should we accept the PR or reject?

**Review by alextp — CHANGES_REQUESTED** [review]
Can you move these changes to assign_* methods on PartitionedVariable, to be more 2.0-friendly?

**Comment by wangsiyu** [comment]
Sorry，I am confused. These assign* methods have already implemented in PartitionedVariable class and I think it convents to 2.0？

**Comment by yangjunpro** [comment]
@alextp 

Hi Alex,

I am also a little bit puzzled as to your comments.

It is appreciated if you could directly comment on our code commit so it would be  easier to ensure we are on the same page.

Thanks.

**Comment by alextp** [comment]
I meant this NotImplementedError: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/variables.py#L2400

If we can replace this error with an implementation of assign, assign_add, etc for PartitionedVariable this will be more 2.0-friendly.

**Review by alextp — DISMISSED** [review]

File: tensorflow/python/ops/variables.py [hunk 1]
```diff
@@ -2395,11 +2395,50 @@ def _get_variable_list(self):
...
-    _ = value, use_locking
-    raise NotImplementedError(
-        "assign() has not been implemented for PartitionedVariable.")
+  def _apply_assign_fn(self,
+                       assign_fn,
+                       value):
+    partition_axes = self._partition_axes()
+    if len(partition_axes) > 1:
+      raise NotImplementedError(
+          "Multi-axis partition assign_fn is not supported "
```
Comment (facaiy) [inline tensorflow/python/ops/variables.py]: missing %s ?

**Comment by wangsiyu** [comment]
@alextp I have refined the coding style according to @facaiy . Please review again. Thanks.

**Comment by wangsiyu** [comment]
Any update for me？

**Review by alextp — APPROVED** [review]

**Pull request merged**